source/grid/manifold_lib.cc

```diff
@@ -301,68 +301,148 @@ get_new_point (const std::vector<Point<spacedim> > &vertices,
-                                                       const double tolerance) :
-  direction (direction),
-  point_on_axis (point_on_axis),
-  tolerance(tolerance)
+CylindricalManifold<dim, spacedim>::CylindricalManifold(const Point<spacedim> &direction_,
+                                                        const Point<spacedim> &point_on_axis_,
+                                                        const double /*tolerance*/) :
+  ChartManifold<dim,3,3>(Tensor<1,3>({0,2.*numbers::PI,0})),
+              normal_direction(compute_normal(direction_)),
+              direction (direction_/direction_.norm()),
+              point_on_axis (point_on_axis_)
 {
-  Assert(spacedim > 2, ExcImpossibleInDim(spacedim));
+  static_assert(spacedim==3,
+                "CylindricalManifold can only be used for spacedim==3!");
 }
 
 
 
+template <int dim, int spacedim>
+Point<spacedim>
+CylindricalManifold<dim, spacedim>::compute_normal(const Tensor<1,spacedim> &vector) const
+{
+  AssertThrow(vector.norm() != 0.,
+              ExcMessage("The direction parameter must not be zero!"));
+  Point<3> normal;
+  if (std::abs(vector[0]) >= std::abs(vector[1])
+      && std::abs(vector[0]) >= std::abs(vector[2]))
+    {
+      normal[1]=-1.;
+      normal[2]=-1.;
+      normal[0]=(vector[1]+direction[2])/vector[0];
+    }
+  else if (std::abs(vector[1]) >= std::abs(vector[0])
+           && std::abs(vector[1]) >= std::abs(vector[2]))
+    {
+      normal[0]=-1.;
+      normal[2]=-1.;
+      normal[1]=(vector[0]+vector[2])/vector[1];
+    }
+  else
+    {
+      normal[0]=-1.;
+      normal[1]=-1.;
+      normal[2]=(vector[0]+vector[1])/vector[2];
+    }
+  return normal;
+}
+
+
 
 template <int dim, int spacedim>
 Point<spacedim>
 CylindricalManifold<dim,spacedim>::
 get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
                const std::vector<double>           &weights) const
 {
-  // compute a proposed new point
-  const Point<spacedim> middle = flat_manifold.get_new_point(surrounding_points,
-                                                             weights);
+  // First check if the the average in space lies on the axis.
+  Point<spacedim> middle;
+  for (unsigned int i=0; i<surrounding_points.size(); ++i)
+    middle += surrounding_points[i]*weights[i];
 
-  double radius = 0;
-  Tensor<1,spacedim> on_plane;
+  middle -= point_on_axis;
+  const double lambda = middle*direction;
 
-  for (unsigned int i=0; i<surrounding_points.size(); ++i)
-    {
-      on_plane = surrounding_points[i]-point_on_axis;
-      on_plane = on_plane - (on_plane*direction) * direction;
-      radius += weights[i]*on_plane.norm();
-    }
+  if ((middle-direction*lambda).square() < 1.e-10*surrounding_points[0].square())
+    return Point<spacedim>()+direction*lambda;
+  else
+    return ChartManifold<dim, spacedim, 3>::get_new_point(surrounding_points, weights);
```

Comment:
Why do you need to treat the axis point separately? I think the manifold doesn't work on the axis points and is invalid anyway, isn't it?

Are there any tests that run into this?